### DISCUSSION TITLE
Welcome to nginx-proxy-manager Discussions!

### DISCUSSION THREAD

**Discussion by jc21** [NginxProxyManager/nginx-proxy-manager#1201] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jiriteach** [comment]
Wahoo! - Thanks for enabling this. Provides an easier forum for many to ask questions. 

Thanks to everyone involved in this project! I use NPM as a core for my home network. Really appreciate the effort that everyone has put into making NPM a success. Keep up the great work all :)

I've learnt lots about Docker and NPM as a result of using for a long while now. 

Thanks
J - From New Zealand

**Comment by AznRaven** [comment]
how do we redirect all sites from http to https? or do we have to do it one site at a time?

**Reply by jiriteach** [reply]
I assume you are referring to a wildcard domain and you would need a wildcard SSL certifcate as well. 
Some commentary here - https://github.com/jc21/nginx-proxy-manager/issues/749.

**Reply by AznRaven** [reply]
I have a synology nas that will handle all the ssl certs. but it's not redirecting http to https and it seems like it's something that can't be done without jumping through hoops. 

Is NPM able to redirect all http traffic to https? Right now, I'm redirecting 1 source to 1 destination in the Redirection Hosts and it works. 
http --> a.domain.com --> 301 --> https --> a.domain.com
http --> b.domain.com --> 301 --> https --> b.domain.com

But what I want is a single redirect command that will redirect all 
http --> *.domain.com --> 301 --> https --> *.domain

**Reply by jiriteach** [reply]
Synology NAS’s can do that. Just use the in built reverse proxy to do the redirects or there is an option to sent to HTTPS. I’ve done it before but it’s not as powerful as NPM.

**Reply by AznRaven** [reply]
did u redirect individual http to https? or all http to https? and how was it done if you don't mind telling? because i couldn't get it. if you google synology http to https online... it seems like a everyone is having problems redirecting http to https on synology

**Reply by AznRaven** [reply]
i couldn't find a working way of doing it

**Comment by casp3yrulezz** [comment]
Hi Guys,,

this is great tool. 
i have setup in my lab environment, i have observed, i have setup domain as proxy host only for https  but its working http. my web server is allowed http. so can you guys please guide me how can i block or control port 80 from the NPM ?? any learning book or anything you guys have please share.

**Reply by chaptergy** [reply]
You have two options. You could either just enable HSTS in a hosts settings, which would leave http theoretically still available, but all modern browsers will block the connection via http, and only allow https traffic.  
The other option is to remove the port mapping  for port 80 from your docker-compose. That way port 80 is never exposed to the outside. This however also breaks the Letsencrypt HTTP challenge, meaning you can only issue SSL certificates using the DNS challenge.

**Reply by casp3yrulezz** [reply]
Hi Chaptergy 

thanks for reply ,, i am aware of second option, so that option its not working for me, first option i am looking, i have IIS web server so can you guide me how can i do ??
i have multiple sites on single servers.no any other option other than this two ?? like can i not apply any of the configurtiaon in Advance configuration ??

**Reply by chaptergy** [reply]
Unfortunately I am not familiar at all with an IIS web server, so you would have to find that infomation for yourself.  
But in NPM itself it is possible by just edititng a host and flipping a switch. The "Force SSL" switch will rewrite all requests via http to https on the nginx proxy side, while the HSTS will tell the browsers to never allow a non-ssl-secured connection to this webpage.
![image](https://user-images.githubusercontent.com/26956711/127995490-8dc98584-09f5-4632-b9bf-0b0d1f5faad3.png)

**Reply by jiriteach** [reply]
@casp3yrulezz - Bit confused as to what you are trying to achieve. You mention IIS so you are hosting multiple sites on the same IIS server and are using host headers all coming in on 443 I assume yes? Then you have NPM in front of this and its working for 443 but also 80 and you want to block 80?

Can you explain further please?

**Reply by casp3yrulezz** [reply]
@Jiriteach yes your are correct ,, any request come for any of the sites for port 80 NPM need to block or redirect any other page for the security reason, only 443 is accessible is required. as earlier given two  solution by @chaptergy but i dnt want to apply second solution because let's encrypt will not work. and for first option i cnt do entry for multiple sites each and every time,, so any setting which i apply during setup domain on NPM ??

**Reply by jiriteach** [reply]
Given that NPM will be your entry point - utlimately it needs to have the certs in place for each domain that needs to be forwarded.

You could use 1 entry with multiple hostnames entered which force's ssl as @chaptergy has suggested but you will have to have a wildcard cert or similar if this scenario supports it.

If you have different domains - then you will have to have the certs setup in NPM and then setup NPM proxy rules for each.

**Reply by casp3yrulezz** [reply]
@Jiriteach what NPM proxy rules need to apply and how its work.

**Reply by casp3yrulezz** [reply]
@chaptergy  is working now.

**Reply by jiriteach** [reply]
> @Jiriteach what NPM proxy rules need to apply and how its work.

See the NPM documentation

**Comment by Morfild** [comment]
Hello everyone, 

I would like to know, is there some file or API I can use to add a new proxy?
I'm developing a service, where users can add their own domain, so I need somehow to register it in the settings from the code.
Or maybe I do something wrong, I created a special domain let's say blabla.domain.com and I give this domain to the customers, they have to register their domain with CNAM to 'blabla.domain.com', but I still have to register their domains in Nginx Proxy Manager, maybe there is some option to allow all CNAME's to proxy?

**Reply by chaptergy** [reply]
There is an api, which is documented but the documentation could be slightly outdated. See https://github.com/jc21/nginx-proxy-manager/issues/341#issuecomment-839129341.

**Reply by Morfild** [reply]
I see there is a few POST endpoints, but unrelated to create a new proxy, or am I wrong?

**Reply by chaptergy** [reply]
Yeah, it is not included in the swagger file. But have a look at [this json schema](https://gist.github.com/chaptergy/a8b28330b74159355f5ecc5ffa488b17#file-npm-json-schema-json-L2257), it should give you an idea of what you have to send to the backend for it to work. 
I've always found it easiest to open the developer console in my browser and switch to the network tab in there. Then I can just execute what I want to do in the frontend, and see exactly what request was send with what data to the backend, so I can replicate the behaviour.

**Comment by casp3yrulezz** [comment]
Does NPM is working as Load Balancing ?? for example NPM is my front end and back end i have two web servers,, single domain can i pointing to  two different servers as a load balancing  ??

**Reply by chaptergy** [reply]
It does not, see https://github.com/jc21/nginx-proxy-manager/issues/156.
